sockeye/training.py

```diff
@@ -328,9 +337,28 @@ def _determine_improvement(self, val_metrics: List[loss.LossMetric]) -> bool:
         for val_metric in val_metrics:
             if val_metric.name == self.config.early_stopping_metric:
                 value = val_metric.get()
-                if utils.metric_value_is_better(value,
-                                                self.state.best_metric,
-                                                self.config.early_stopping_metric):
+                # When using Horovod, the primary worker makes an authoritative
+                # check of whether metric value has improved and broadcasts the
+                # result to secondary workers.  Non-determinism in the order of
+                # GPU operations can lead to slight numeric variation across
+                # workers, causing potential desync if each worker makes its own
+                # check for key training decisions (reducing learning rate,
+                # early stopping, etc.).
+                if self.config.using_horovod and horovod_mpi.hvd.rank() > 0:
+                    # Horovod secondary workers: wait for primary worker to send
+                    # result.
+                    value_is_better = None  # type: bool
```

Comment:
shouldn't this be `# type: Optional[bool]`?

Reply:
Good catch